### PR TITLE
Forum Categories having colors and edit category features

### PR DESCRIPTION
**Pull request by scopeInfinity**
Issue: #2094 (Forum: color coded & instructor ordered tags)

To be merged after PR Forum Edit Categories (Reordering) #2126, need to change Base Branch of PR before merging.

Tasks
- [x] Categories with color tags
- [x] Modify color tags
- [x] Edit Category names
- [x] After adding category, it should be added in cat-buttons
- [x] Remove Category
- [ ] ~Edit Thread Title~
- [ ] ~Allow Category of existing threads to be modified~
- [x] Dynamically refreshing categories buttons on Create Thread page(for implemented features)
- [ ] ~Instructor/full access ta should be able to view deleted threads~

### CODE REVIEW

File: .setup/update_database.py [hunk 1]
```diff
@@ -133,6 +133,7 @@
         
         # categories ordering for discussion forum
         os.system("PGPASSWORD='{}' psql --host={} --username={} --dbname={} -c 'ALTER TABLE ONLY categories_list ADD COLUMN rank int'".format(*variables))
+        os.system("PGPASSWORD='{}' psql --host={} --username={} --dbname={} -c 'ALTER TABLE ONLY categories_list ADD COLUMN color varchar DEFAULT '#000080'".format(*variables))
```
Comment (andrewaikens87) [inline .setup/update_database.py]: Please test this, there is a single quote missing at the end to close off the one before "ALTER"

File: site/app/controllers/forum/ForumController.php [hunk 2]
```diff
@@ -172,6 +189,67 @@ public function addNewCategory(){
...
+
+    public function editCategory(){
+        $result = array();
+        if($this->core->getUser()->getGroup() <= 2){
+            $category_id = $_REQUEST["category_id"];
+            $category_desc = null;
+            $category_color = null;
+            $should_update = true;
+
+            if((!empty($_REQUEST["category_desc"])) && isset($_REQUEST["category_desc"])) {
```
Comment (andrewaikens87) [inline site/app/controllers/forum/ForumController.php]: Take a look at [https://www.codepunker.com/blog/is-null-vs-empty-vs-isset-one-lesson-all-php-coders-should-learn](https://www.codepunker.com/blog/is-null-vs-empty-vs-isset-one-lesson-all-php-coders-should-learn)

Also refer to [http://php.net/manual/en/function.isset.php](http://php.net/manual/en/function.isset.php) and [http://php.net/manual/en/function.empty.php](http://php.net/manual/en/function.empty.php)

Comment (scopeInfinity) [reply site/app/controllers/forum/ForumController.php]: Thanks! I didn't know that.

**Comment by andrewaikens87** [comment]
Can we display the create category button inline?
screenshot from 2018-06-14 00-43-03

**Comment by andrewaikens87** [comment]
Other than the small tweaks I would say the behavior is there and we can modify the ui as needed.

**Comment by saxocellphone** [comment]
I'm getting a JS error data.trim is not a function in server.js:1420:9 when trying to create a thread as a student. 
capture

**Comment by saxocellphone** [comment]
Also I think it will be nice if the newly added categories show up on the list immediately after being added, instead of closing and reopening

File: migration/data/course_tables.sql [hunk 3]
```diff
@@ -515,6 +515,8 @@ CREATE TABLE "thread_categories" (
 CREATE TABLE "categories_list" (
 	"category_id" serial NOT NULL,
 	"category_desc" varchar NOT NULL,
+	"rank" int,
+	"color" varchar DEFAULT '#000080',
```
Comment (MasterOdin) [inline migration/data/course_tables.sql]: can this be null?

Comment (scopeInfinity) [reply migration/data/course_tables.sql]: No, it can't be NULL. Thankyou!

**Comment by scopeInfinity** [comment]
@andrewaikens87 Yah, it wasn't visibly populating categories for students, indeed a big bug.. Thankyou!
Fixed it.

**Comment by MasterOdin** [comment]
Are we fine with this using the update_database.py or should we add a migration for this?

**Comment by bmcutler** [comment]
Shoot, I forgot to check for database changes.
No, we want to undo the edits to update_database and make it a migration.
Can you take care of that?  Just make a new PR with the edits
(or we could revert...)

**Comment by MasterOdin** [comment]
I'll make a new PR.

**Pull request merged**